impeller/entity/contents/filters/gaussian_blur_filter_contents.cc

```diff
@@ -210,30 +210,38 @@ Entity ApplyClippedBlurStyle(Entity::ClipOperation clip_operation,
                              const Snapshot& input_snapshot,
                              Entity blur_entity,
                              const std::shared_ptr<Geometry>& geometry) {
-  auto shared_blur_entity = std::make_shared<Entity>(std::move(blur_entity));
-  shared_blur_entity->SetNewClipDepth(entity.GetNewClipDepth());
-  auto clipper = std::make_unique<ClipContents>();
-  clipper->SetClipOperation(clip_operation);
-  clipper->SetGeometry(geometry);
+  blur_entity.SetNewClipDepth(entity.GetNewClipDepth());
```

Comment:
This line needs to happen inside the anonymous contents. When FilterContents is rendering the final Entity to the pass, the "new" depth gets overridden with the correct depth just before rendering.

Reply:
So, that fixes the case for the gradient.  Solid colors are broken though still when I turn stencil and cover on.

Reply:
(I just pushed up that change)

Reply:
Okay, I got this sorted out.  It renders correctly now.  It's a bit surprising to me what I needed to do to get Solid working.  I'm not 100% convinced this is safe to do though since calling render twice could result in different results because its modifying the captured entity =T.

Reply:
Okay, i made the functions idempotent so it should be safe.  Here is the current state:

| | solid | inner | outer |
|-|-|-|-|
| **solid color** |✅ |✅ |✅|
| **gradient** | works without stc  |✅ |✅|

I'm going to need brandon's help trying to get that last one solved.  The 2 entities render fine independently, trying to draw them together ends up with one of them clipped.  I don't know how to debug this.

Reply:
Okay I got this to work by just drawing the original piece first.  I know the stencil is going to cut it out anyway.  I thought I understand what was going on but I guess not.  I filed a documentation request and scheduled some time time with bdero ( https://github.com/flutter/flutter/issues/144943 )

Reply:
I pulled down 25b61b996f910bc35a7a78cac604df228f5267ff (latest commit in your branch at the time of writing) for a look and found several places where the new clip depth isn't getting set. The diff rendering on GitHub doesn't appear to match 25b61b996f910bc35a7a78cac604df228f5267ff, so here's the diff that makes everything work:

```diff
diff --git a/impeller/entity/contents/filters/gaussian_blur_filter_contents.cc b/impeller/entity/contents/filters/gaussian_blur_filter_contents.cc
index eab1356af6..2c1f4f5077 100644
--- a/impeller/entity/contents/filters/gaussian_blur_filter_contents.cc
+++ b/impeller/entity/contents/filters/gaussian_blur_filter_contents.cc
@@ -210,7 +210,6 @@ Entity ApplyClippedBlurStyle(Entity::ClipOperation clip_operation,
                              const Snapshot& input_snapshot,
                              Entity blur_entity,
                              const std::shared_ptr<Geometry>& geometry) {
-  blur_entity.SetNewClipDepth(entity.GetNewClipDepth());
   auto clip_contents = std::make_shared<ClipContents>();
   clip_contents->SetClipOperation(clip_operation);
   clip_contents->SetGeometry(geometry);
@@ -226,8 +225,10 @@ Entity ApplyClippedBlurStyle(Entity::ClipOperation clip_operation,
                             const ContentContext& renderer,
                             const Entity& entity, RenderPass& pass) mutable {
         bool result = true;
+        clipper.SetNewClipDepth(entity.GetNewClipDepth());
         clipper.SetTransform(entity.GetTransform() * clipper.GetTransform());
         result = clipper.Render(renderer, pass) && result;
+        blur_entity.SetNewClipDepth(entity.GetNewClipDepth());
         blur_entity.SetTransform(entity.GetTransform() *
                                  blur_entity.GetTransform());
         result = blur_entity.Render(renderer, pass) && result;
@@ -277,6 +278,8 @@ Entity ApplyBlurStyle(FilterContents::BlurStyle blur_style,
                                 const ContentContext& renderer,
                                 const Entity& entity,
                                 RenderPass& pass) mutable {
+            blurred.SetNewClipDepth(entity.GetNewClipDepth());
+            snapshot_entity.SetNewClipDepth(entity.GetNewClipDepth());
             return blurred.Render(renderer, pass) &&
                    snapshot_entity.Render(renderer, pass);
           }),
```

Modified test to demonstrate (top to bottom -- Normal, Inner, Outer, Solid):
![Screenshot 2024-03-11 at 2 00 16 PM](https://github.com/flutter/engine/assets/919017/96b7d84b-8fd8-4fb5-ba43-a96da3b663bf)

```cpp
TEST_P(AiksTest, GaussianBlurStyleSolid) {
  Canvas canvas;
  canvas.Scale(GetContentScale());

  canvas.DrawPaint({.color = Color(0.1, 0.1, 0.1, 1.0)});

  auto path = PathBuilder()
                  .MoveTo({200, 200})
                  .LineTo({300, 400})
                  .LineTo({100, 400})
                  .Close()
                  .TakePath();

  std::vector<FilterContents::BlurStyle> styles = {
      FilterContents::BlurStyle::kNormal, FilterContents::BlurStyle::kInner,
      FilterContents::BlurStyle::kOuter, FilterContents::BlurStyle::kSolid};

  Paint paint;
  paint.color = Color::Green();

  for (auto style : styles) {
    paint.mask_blur_descriptor = Paint::MaskBlurDescriptor{
        .style = style,
        .sigma = Sigma(30),
    };
    canvas.Save();
    {
      canvas.DrawPath(path, paint);
      canvas.Translate({200, 0, 0});
      canvas.DrawPath(path, paint);
    }
    canvas.Restore();
    canvas.Translate({0, 100, 0});
  }

  ASSERT_TRUE(OpenPlaygroundHere(canvas.EndRecordingAsPicture()));
}
```

Reply:
Oops, actually my stage is a bit screwed up it seems. Let me fix that and check again.

Reply:
That doesn't work for me locally.  For `GaussianBlurStyleSolid` the original image is getting clipped by the outer blur draw.  If I swap the draw order it's fine, but not technically as good since its a slight deviation from what we had.

Reply:
I cleared up my stage to get to actual code you're working against and figured out what's going on. In a frame capture I noticed that kSolid was inserting a difference clip in a frame capture by grabbing the outer clipped entity via `ApplyClippedBlurStyle`. Does this diff work for you?

```diff
diff --git a/impeller/entity/contents/filters/gaussian_blur_filter_contents.cc b/impeller/entity/contents/filters/gaussian_blur_filter_contents.cc
index 2c177141c7..98535695fc 100644
--- a/impeller/entity/contents/filters/gaussian_blur_filter_contents.cc
+++ b/impeller/entity/contents/filters/gaussian_blur_filter_contents.cc
@@ -228,6 +228,7 @@ Entity ApplyClippedBlurStyle(Entity::ClipOperation clip_operation,
         clipper.SetTransform(entity.GetTransform() * entity_transform);
         result = clipper.Render(renderer, pass) && result;
         blur_entity.SetClipDepth(clipper.GetClipDepth());
+        blur_entity.SetNewClipDepth(entity.GetNewClipDepth());
         blur_entity.SetTransform(entity.GetTransform() * blur_transform);
         result = blur_entity.Render(renderer, pass) && result;
         if constexpr (!ContentContext::kEnableStencilThenCover) {
@@ -265,31 +266,28 @@ Entity ApplyBlurStyle(FilterContents::BlurStyle blur_style,
                                    input, input_snapshot,
                                    std::move(blur_entity), geometry);
     case FilterContents::BlurStyle::kSolid: {
-      Entity blurred = ApplyClippedBlurStyle(Entity::ClipOperation::kDifference,
-                                             entity, input, input_snapshot,
-                                             std::move(blur_entity), geometry);
       Entity snapshot_entity = Entity::FromSnapshot(
           input_snapshot, entity.GetBlendMode(), entity.GetClipDepth());
       Entity result;
-      Matrix blurred_transform = blurred.GetTransform();
+      Matrix blurred_transform = blur_entity.GetTransform();
       Matrix snapshot_transform = snapshot_entity.GetTransform();
       result.SetContents(Contents::MakeAnonymous(
-          fml::MakeCopyable(
-              [blurred = blurred.Clone(), blurred_transform, snapshot_transform,
-               snapshot_entity = std::move(snapshot_entity)](
-                  const ContentContext& renderer, const Entity& entity,
-                  RenderPass& pass) mutable {
-                bool result = true;
-                snapshot_entity.SetTransform(entity.GetTransform() *
-                                             snapshot_transform);
-                snapshot_entity.SetNewClipDepth(entity.GetNewClipDepth());
-                result = result && snapshot_entity.Render(renderer, pass);
-                result = result && snapshot_entity.Render(renderer, pass);
-                blurred.SetClipDepth(snapshot_entity.GetClipDepth());
-                blurred.SetTransform(entity.GetTransform() * blurred_transform);
-                result = result && blurred.Render(renderer, pass);
-                return result;
-              }),
-          fml::MakeCopyable([blurred = blurred.Clone(),
+          fml::MakeCopyable([blurred = blur_entity.Clone(), blurred_transform,
+                             snapshot_transform,
+                             snapshot_entity = std::move(snapshot_entity)](
+                                const ContentContext& renderer,
+                                const Entity& entity,
+                                RenderPass& pass) mutable {
+            blurred.SetNewClipDepth(entity.GetNewClipDepth());
+            blurred.SetTransform(entity.GetTransform() * blurred_transform);
+
+            snapshot_entity.SetNewClipDepth(entity.GetNewClipDepth());
+            snapshot_entity.SetTransform(entity.GetTransform() *
+                                         snapshot_transform);
+            return blurred.Render(renderer, pass) &&
+                   snapshot_entity.Render(renderer, pass);
+          }),
+          fml::MakeCopyable([blurred = blur_entity.Clone(),
                              blurred_transform](const Entity& entity) mutable {
             blurred.SetTransform(entity.GetTransform() * blurred_transform);
             return blurred.GetCoverage();
```

Reply:
Chatted with brandon.  I was thinking about this all wrong.  Since we are just generating a mask, blending is not a concern so there is no need for clipping in the Solid case, we can just draw the blurred entity and the original entity over top of it.

As to why this didn't work as expected, where it works with the old clip but not the stencil-then-cover, was that there is a limitation to the clip depth with stc that didn't exist with the old system.  That's moot though because of the aforementioned point.

Reply:
Done, thanks @bdero